plugins/builtin/apm/nomad/plugin/job.go

```diff
@@ -0,0 +1,179 @@
+// contains all the information needed to perform a Nomad APM query for a task
+// group.
+type taskGroupQuery struct {
+	metric    string
+	job       string
+	group     string
+	operation string
+}
+
+func (a *APMPlugin) queryTaskGroup(q string) (float64, error) {
+
+	// Parse the query ensuring we have all information available to make all
+	// subsequent calls.
+	query, err := parseTaskGroupQuery(q)
+	if err != nil {
+		return 0, fmt.Errorf("failed to parse query: %v", err)
+	}
+	a.logger.Debug("expanded query", "from", q, "to", fmt.Sprintf("%# v", query))
+
+	metrics, err := a.getTaskGroupResourceUsage(query)
+	if err != nil {
+		return 0, err
+	}
+
+	if len(metrics) == 0 {
+		return 0, fmt.Errorf("metric not found: %s", q)
+	}
+	a.logger.Debug("metrics found", "num_data_points", len(metrics), "query", q)
+
+	return calculateTaskGroupResult(query.operation, metrics), nil
+}
+
+// getTaskGroupResourceUsage iterates the allocations within a job and
+// identifies those which meet the criteria for being part of the calculation.
+func (a *APMPlugin) getTaskGroupResourceUsage(query *taskGroupQuery) ([]float64, error) {
+
+	// Grab the list of allocations assigned to the job in question.
+	allocs, _, err := a.client.Jobs().Allocations(query.job, false, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get alloc listing for job: %v", err)
+	}
+
+	// The response is a list of data points from each allocation running in
+	// the task group.
+	var resp []float64
+
+	// Define a function that manages updating our response.
+	metricFunc := func(m *[]float64, ru *api.ResourceUsage) {}
+
+	// Depending on the desired metric, the function will append different data
+	// to the response. Using a function means we only have to perform the
+	// switch a single time, rather than on a per allocation basis.
+	switch query.metric {
+	case queryMetricCPU:
+		metricFunc = func(m *[]float64, ru *api.ResourceUsage) {
+			*m = append(*m, ru.CpuStats.Percent)
+		}
+	case queryMetricMem:
+		metricFunc = func(m *[]float64, ru *api.ResourceUsage) {
+			*m = append(*m, float64(ru.MemoryStats.Usage))
+		}
+	}
+
+	for _, alloc := range allocs {
+
+		// If the allocation is not running, or is not part of the target task
+		// group then we should skip and move onto the next allocation.
+		if alloc.ClientStatus != api.AllocClientStatusRunning || alloc.TaskGroup != query.group {
+			continue
+		}
+
+		// Obtains the statistics for the task group allocation. If we get a
+		// single error during the iteration, we cannot reliably make a scaling
+		// calculation.
+		//
+		// When calling Stats an entire Allocation object is needed, but only
+		// the ID is used within the call. Further details:
+		// https://github.com/hashicorp/nomad/issues/7955
+		allocStats, err := a.client.Allocations().Stats(&api.Allocation{ID: alloc.ID}, nil)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get alloc stats: %v", err)
+		}
+
+		// Be safe, be sensible.
```

Comment:
above, "a single error [means] we cannot reliably make a scaling calculation"
what about a single missing allocation status like here?

Reply:
So thinking back to the logic here I am expecting that an error indicates we were not able to call the API, but checking for a nil alloc is incase the allocID is returned in the list but then is GC'd.

It may be that an alloc that gets GC'd might return an error with 404 in the string rather than a nil alloc in the same way I know jobs behaves. I'll raise an issue to check this logic; its only showing up in this commit as its moved file.

Reply:
👍